addons/mail/models/mail_message.py

```diff
@@ -941,92 +941,68 @@ def _notify_moderators(self):
+            if message_sudo.author_id:
+                author = message_sudo.author_id.name_get()[0]
             else:
-                author = (0, message.email_from)
+                author = (0, message_sudo.email_from)
+
+            # Notifications
             customer_email_status = (
-                (all(n.notification_status == 'sent' for n in message.notification_ids if n.notification_type == 'email') and 'sent') or
-                (any(n.notification_status == 'exception' for n in message.notification_ids if n.notification_type == 'email') and 'exception') or
-                (any(n.notification_status == 'bounce' for n in message.notification_ids if n.notification_type == 'email') and 'bounce') or
+                (all(n.notification_status == 'sent' for n in message_sudo.notification_ids if n.notification_type == 'email') and 'sent') or
+                (any(n.notification_status == 'exception' for n in message_sudo.notification_ids if n.notification_type == 'email') and 'exception') or
+                (any(n.notification_status == 'bounce' for n in message_sudo.notification_ids if n.notification_type == 'email') and 'bounce') or
                 'ready'
             )
             customer_email_data = []
-            for notification in email_notification_tree[message.id]:
-                customer_email_data.append((partner_tree[notification.res_partner_id.id][0], partner_tree[notification.res_partner_id.id][1], notification.notification_status))
+            filtered_notifications = message_sudo.notification_ids.filtered(lambda n:
+                n.notification_type == 'email' and n.res_partner_id.active and
+                (n.notification_status in ('bounce', 'exception', 'canceled') or n.res_partner_id.partner_share)
+            )
+            for notification in filtered_notifications:
+                partner_name_get = notification.res_partner_id.name_get()[0]
+                customer_email_data.append((partner_name_get[0], partner_name_get[1], notification.notification_status))
+
+            has_access_to_model = message_sudo.model and self.env[message_sudo.model].check_access_rights('read', raise_exception=False)
 
-            has_access_to_model = message.model and self.env[message.model].check_access_rights('read', raise_exception=False)
-            main_attachment = None
-            if message.attachment_ids and message.res_id and issubclass(self.pool[message.model], self.pool['mail.thread']) and has_access_to_model:
-                main_attachment =  self.env[message.model].browse(message.res_id).message_main_attachment_id
+            # Attachments
+            main_attachment = self.env['ir.attachment']
+            if message_sudo.attachment_ids and message_sudo.res_id and issubclass(self.pool[message_sudo.model], self.pool['mail.thread']) and has_access_to_model:
+                main_attachment = self.env[message_sudo.model].browse(message_sudo.res_id).message_main_attachment_id
```

Comment:
What is the purpose of has_access_to_model if you are reading message_main_attachment_id afterwards ? Shoudln't we try / except it instead ?

Reply:
This was the old code, I didn't change the logic here. I'll double check if it is still necessary, and either add a comment with my finding or remove it.

Reply:
I'm going to remove this condition and do it as sudo:
- from portal everything is sudo so it's pointless to check access on top of it
- from backend, it's almost impossible to trigger the case, and it's not like returning is_main True would leak any sensitive information to an employee